addons/website_sale_referral/models/referral_mixin.py

```diff
@@ -0,0 +1,82 @@
+
+
+class ReferralMixin(models.AbstractModel):
+    """ Mixin class for objects which can be tracked by referral. """
+    _name = 'referral.mixin'
+    _description = 'Referral Mixin'
+
+    STATES_PRIORITY = {'cancel': 0, 'new': 1, 'in_progress': 2, 'done': 3}
+    REFERRAL_STAGES = {'new': 'New', 'in_progress': 'In Progress', 'done': 'Done'}
+
+    to_reward = fields.Boolean()
+    reward_done = fields.Boolean()
+
+    def get_referral_statuses(self, utm_source_id, referred_email=None):
+        objects = self.find_others(utm_source_id, referred_email)
+
+        result = {}
+        for o in objects:
+            state = o._get_state_for_referral()
+            if(o.referred_email not in result or self.STATES_PRIORITY[state] > self.STATES_PRIORITY[result[o.referred_email]['state']]):
+                result[o.referred_email] = {'state': state, 'name': o.referred_name, 'company': o.referred_company}
+
+        if referred_email:
+            return result.get(referred_email, None)
+        else:
+            return result
+
+    def check_referral_progress(self, old_state, new_state):
+        others_to_reward = self.find_others(self.source_id, referred_email=self.referred_email, extra_criteria=[('to_reward', '=', True)])
+        if new_state != old_state and not len(others_to_reward):
+            self.get_referral_tracking().updates_count += 1
+            if new_state == 'done':
+                self._send_mail('referral_won_email_template', 'Referral won !', {'referred_name': self.referred_name})
+
+                responsible_id = self.env['ir.config_parameter'].sudo().get_param('website_sale_referral.responsible_id') or SUPERUSER_ID
+                if responsible_id:
+                    activity = self.activity_schedule(
+                        act_type_xmlid='website_sale_referral.mail_act_data_referral_reward',
+                        summary='The referrer for this lead deserves a reward',
+                        user_id=responsible_id)
+                    if(callable(getattr(self, 'activity_created', None))):
+                        self.activity_created(activity)
+                    self.to_reward = True
+
+            elif new_state == 'cancel':
+                self._send_mail('referral_cancelled_email_template', 'Referral lost...', {'referred_name': self.referred_name})
+            else:
+                self._send_mail('referral_state_changed_email_template', 'Referral progressed !', {'referred_name': self.referred_name, 'state': _(self.REFERRAL_STAGES[new_state])})
+
+    def find_others(self, utm_source_id, referred_email=None, extra_criteria=[]):
+        criteria = [
+            ('campaign_id', '=', self.env.ref('website_sale_referral.utm_campaign_referral').id),
+            ('source_id', '=', utm_source_id.id)]
+        if(referred_email):
+            criteria.append(('referred_email', '=', referred_email))
+        if(extra_criteria):
+            criteria.extend(extra_criteria)
```

Comment:
By safer, I mean that here we could pass and arbitrary domain, that is directly added to your domain. And executed in sudo...

So it's not very difficult to forge something nasty.

Btw, you may be interested into this presentation too ;)
https://docs.google.com/presentation/d/1oDINxPtHWz31V8-2W0h2u2ubaKgz9lmbyfx9DJI4lTw/edit#slide=id.g2bac39f39e_0_0

Reply:
Done